### PR TITLE
Verity oauth token origin

### PR DESCRIPTION
**Pull request by vgarvardt**

### CODE REVIEW

File: cmd/janus/main.go [hunk 1]
```diff
@@ -68,8 +68,12 @@ func main() {
 	r.Use(middleware.NewLogger(globalConfig.Debug).Handler, middleware.NewRecovery(RecoveryHandler).Handler, middleware.NewMongoDB(accessor).Handler)
 
 	// create the proxy
+	oAuthServersRepo, err := oauth.NewMongoRepository(accessor.Session.DB(""))
+	if err != nil {
+		panic(err)
```
Comment (italolelis) [inline cmd/janus/main.go]: Please use `log.Panic`

File: pkg/oauth/key_exists_middleware.go [hunk 2]
```diff
@@ -62,6 +63,17 @@ func (m *KeyExistsMiddleware) Handler(handler http.Handler) http.Handler {
...
 
+		if m.oAuthServerID != thisSessionState.OAuthServerID {
+			log.WithFields(log.Fields{
+				"path":   r.RequestURI,
+				"origin": r.RemoteAddr,
+				"key":    accessToken,
+				"sessionOAuthServerID": thisSessionState.OAuthServerID,
+				"authOAuthServerID":    m.oAuthServerID,
+			}).Info("Attempted access with the key issued by other OAuth provider.")
+			panic(ErrAccessTokenOfOtherOrigin)
```
Comment (italolelis) [inline pkg/oauth/key_exists_middleware.go:75]: Please use `log.Panic`

File: pkg/oauth/mongodb_repository.go [hunk 3]
```diff
@@ -101,3 +101,12 @@ func (r *MongoRepository) Remove(id string) error {
 	log.Debugf("Resource %s removed", id)
 	return nil
 }
+
+// FindOAuthServerByTokenURL returns OAuth server records with corresponding token url
+func (r *MongoRepository) FindOAuthServerByTokenURL(url string) (*OAuth, error) {
```
Comment (italolelis) [inline pkg/oauth/mongodb_repository.go]: Let's call it `FindByTokenURL`

Comment (italolelis) [reply pkg/oauth/mongodb_repository.go]: Also, don't accept a string, let's accept a `url.URL`

File: pkg/oauth/transport.go [hunk 4]
```diff
@@ -75,6 +78,15 @@ func (t *RoundTripper) RoundTrip(req *http.Request) (resp *http.Response, err er
 		bodyBytes, _ = ioutil.ReadAll(resp.Body)
 
 		if marshalErr := json.Unmarshal(bodyBytes, &newSession); marshalErr == nil {
+			tokenURL := url.URL{Scheme: req.URL.Scheme, Host: req.URL.Host, Path: req.URL.Path}
+			log.WithField("token_url", tokenURL.String()).Debug("Looking for OAuth provider who issued the token")
+			oAuthServer, err := t.oAuthServersRepo.FindOAuthServerByTokenURL(tokenURL.String())
+			if err != nil {
+				log.Error("Failed to find OAuth server by token URL", err)
+				return resp, err
```
Comment (italolelis) [inline pkg/oauth/transport.go]: We can't return here if we do the body on line 95 won't be restored.

File: pkg/oauth/key_exists_middleware.go [hunk 5]
```diff
@@ -19,12 +20,13 @@ var (
 
 // KeyExistsMiddleware checks the integrity of the provided OAuth headers
 type KeyExistsMiddleware struct {
-	manager *Manager
+	manager       *Manager
+	oAuthServerID bson.ObjectId
```
Comment (italolelis) [inline pkg/oauth/key_exists_middleware.go:25]: Passing the `api.Spec` pointer is a bit more flexible and probably will allow us to check for other things in the future

**Review by italolelis — CHANGES_REQUESTED** [review]

File: pkg/oauth/key_exists_middleware.go (hunk 2 above)
Comment (vgarvardt) [reply pkg/oauth/key_exists_middleware.go:75]: middleware must panic in normal way

**Comment by italolelis** [comment]
👍 

[Approved with PullApprove](https://pullapprove.com/hellofresh/api-gateway/pull-request/48/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=hellofresh/api-gateway)

**Review by italolelis — APPROVED** [review]

**Pull request merged**